syntax = "proto3";

package heron.proto.tmaster;

option java_package = "com.twitter.heron.proto.tmaster";
option java_outer_classname = "TopologyMaster";

import "common.proto";
import "stats.proto";
import "physical_plan.proto";

message TMasterLocation {
  string topology_name = 1;
  string topology_id = 2;
  string host = 3;

  // The port to talk to the tmaster
  // for topology control actions
  // like Activate/DeActivate/Drain etc
  int32 controller_port = 4;

  // The port that different components
  // of the topology use to talk to the tmaster
  // like stmgr for getting pplans
  // and metrics mgr for sending stats
  int32 master_port = 5;

  // The port that is a http endpoint
  // to publish stats about the topology
  int32 stats_port = 6;
}

//
// interface called by the StreamManagers.
//
message StMgrRegisterRequest {
  // Info about myself.
  heron.proto.system.StMgr stmgr = 1;

  // All the instances that are on my node.
  repeated heron.proto.system.Instance instances = 2;
}
message StMgrRegisterResponse {
  // What was the status of this registration reqeust
  heron.proto.system.Status status = 1;

  // If the assignment is already known, its sent
  heron.proto.system.PhysicalPlan pplan = 2;
}

message StMgrHeartbeatRequest {
  // what time the heartbeat was sent at.
  int64 heartbeat_time = 1;

  // Any stats that nodemanager has
  heron.proto.system.StMgrStats stats = 2;
}
message StMgrHeartbeatResponse {
  // What was the status of this heartbeat reqeust
  heron.proto.system.Status status = 1;
}

//
// interface called by the metric managers
// TODO: Rename this message to something else to avoid confusion with metrics::MetricDatum
//
message MetricDatum {
  // What is the name of the component
  // Like Tail-FlatMap or __stmgr__
  string component_name = 1;

  // In case of a regular instance, it is the instance's
  // instance_id. For stmgr it is the stmgr_id
  string instance_id = 2;
  string name = 3;
  string value = 4;
  int64 timestamp = 5;
}

message TmasterExceptionLog {
  // Source of exception.
  string component_name = 1;

  // Current hostname.
  string hostname = 2;

  // In case of a regular instance, it is the instance's
  // instance_id. For stmgr it is the stmgr_id
  string instance_id = 3;

  // TODO: Use heron.proto.system.metrics.ExceptionData isntead of following 5 fields
  // Stack trace of exception. First two lines of stack trace is used for aggregating exception.
  string stacktrace = 4;

  // Last time the exception occurred in the metrics collection interval
  string lasttime = 5;

  // First time the exception occurred in the metrics collection interval
  string firsttime = 6;

  // Number of time exception occurred in the metrics collection interval
  int32 count = 7;

  // Additional text logged with the exception.
  string logging = 8;
}

message PublishMetrics {
  repeated MetricDatum metrics = 1;
  repeated TmasterExceptionLog exceptions = 2;
}

//
// interface called by the UI
// TMaster exposes a web endpoint
// You send a post request with the protobuf
// as the data. The response will be MetricResponse
//
message MetricInterval {
  int64 start = 1;
  int64 end = 2;
}
message MetricRequest {
  string component_name = 1;
  // The instance ids to get the stats from
  // If nothing is specified, we will get from
  // all the instances of the component name
  repeated string instance_id = 2;

  // What set of metrics you are interested in
  // Example is __emit-count/default
  repeated string metric = 3;

  // what timeframe data in seconds
  // -1/0 means everything
  // What is the time interval that you want the metrics for
  // Clients can specify one in many ways.
  // 1. interval. Essentially seconds worth of metrics before the current time
  // 2. explicit_interval. An explicit interval
  int64 interval = 4;
  MetricInterval explicit_interval = 5;

  // Do you want metrics broken down on a per minute basis? Set to false by default
  bool minutely = 6;
}

message MetricResponse {
  message IndividualMetric {
    string name = 1;
    string value = 2;
    message IntervalValue {
      string value = 1;
      MetricInterval interval = 2;
    }
    repeated IntervalValue interval_values = 3;
  }
  message TaskMetric {
    string instance_id = 1;
    repeated IndividualMetric metric = 2;
  }
  heron.proto.system.Status status = 1;

  // The order is the same as the request
  repeated TaskMetric metric = 2;

  // The interval we got the data for.
  // This may not be present in case status is NOTOK
  int64 interval = 3;
}

// List of Exception bucketed by metrics aggregation granularity(minute).
message ExceptionLogResponse {
  heron.proto.system.Status status = 1;

  // List of exceptions.
  repeated TmasterExceptionLog exceptions = 2;
}

// Request to fetch exception from tmaster store of exception and metrics
message ExceptionLogRequest {
  // TODO: Make this repeated so that single request can send data for multiple components.
  string component_name = 1;
  // List of instances for exception request. If instances are empty, exception log
  // for all instances will be returned.
  repeated string instances = 2;
}
